planning.py

```diff
@@ -44,174 +90,169 @@ class Action:
     action is an Expr where variables are given as arguments(args).
     Precondition and effect are both lists with positive and negated literals.
     Example:
-    precond_pos = [expr("Human(person)"), expr("Hungry(Person)")]
-    precond_neg = [expr("Eaten(food)")]
-    effect_add = [expr("Eaten(food)")]
-    effect_rem = [expr("Hungry(person)")]
-    eat = Action(expr("Eat(person, food)"), [precond_pos, precond_neg], [effect_add, effect_rem])
+    precond = [expr("Human(person)"), expr("Hungry(Person)"), expr("~Eaten(food)")]
+    effect = [expr("Eaten(food)"), expr("~Hungry(person)")]
+    eat = Action(expr("Eat(person, food)"), precond, effect)
     """
 
-    def __init__(self, action, precond, effect):
-        self.name = action.op
-        self.args = action.args
-        self.precond_pos = precond[0]
-        self.precond_neg = precond[1]
-        self.effect_add = effect[0]
-        self.effect_rem = effect[1]
-
-    def __call__(self, kb, args):
-        return self.act(kb, args)
-
-    def substitute(self, e, args):
-        """Replaces variables in expression with their respective Propositional symbol"""
-        new_args = list(e.args)
-        for num, x in enumerate(e.args):
-            for i, _ in enumerate(self.args):
-                if self.args[i] == x:
-                    new_args[num] = args[i]
+    def __init__(self, expression, precond, effect):
+        self.name = expression.op
+        self.args = expression.args
+        self.subst = None
+
+        def is_negative_clause(e):
+            return e.op == '~' and len(e.args) == 1
+
+        precond_neg, precond_pos = partition(precond, is_negative_clause)
+        self.precond_pos = set(precond_pos)
+        self.precond_neg = set(e.args[0] for e in precond_neg)  # change the negative Exprs to positive
+        effect_rem, effect_add = partition(effect, is_negative_clause)
+        self.effect_add = set(effect_add)
+        self.effect_rem = set(e.args[0] for e in effect_rem)  # change the negative Exprs to positive
+
+    def __repr__(self):
+        return '{}({}, {}, {})'.format(self.__class__.__name__, Expr(self.name, self.args),
+                                       list(self.precond_pos) + ['~{0}'.format(p) for p in self.precond_neg],
+                                       list(self.effect_add) + ['~{0}'.format(e) for e in self.effect_rem])
+
+    def substitute(self, subst, e):
+        """Replaces variables in expression with the same substitution used for the precondition. """
```

Comment:
A minor thing: you accidentally added a space after the last period. I would also remove the period itself, since it is just a one-line sentence.